File: agentaid-claude-service/services/followupService.js
import Anthropic from '@anthropic-ai/sdk';
import dotenv from 'dotenv';

dotenv.config();

const anthropic = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY,
});

// Check if information is complete and specific enough
export function checkCompleteness(extractedData) {
  const issues = [];
  
  // Check 1: Vague items (medicine, food, supplies)
  const vagueItems = ['medicine', 'medical', 'supplies', 'food', 'items', 'stuff', 'things'];
  const needsSpecificity = extractedData.items.some(item => 
    vagueItems.some(vague => item.toLowerCase().includes(vague))
  );
  
  if (needsSpecificity) {
    issues.push({
      type: 'vague_items',
      field: 'items',
      current_value: extractedData.items,
      question: generateItemSpecificityQuestion(extractedData.items)
    });
  }
  
  // Check 2: Missing or vague quantity
  if (!extractedData.quantity_needed || 
      (typeof extractedData.quantity_needed === 'string' && 
       ['low', 'medium', 'high'].includes(extractedData.quantity_needed.toLowerCase()))) {
    issues.push({
      type: 'vague_quantity',
      field: 'quantity_needed',
      current_value: extractedData.quantity_needed,
      question: `Please specify the exact quantity needed for ${extractedData.items.join(', ')}. For example: "50 units" or "100 bottles"`
    });
  }
  
  // Check 3: Missing contact information
  if (!extractedData.contact) {
    issues.push({
      type: 'missing_contact',
      field: 'contact',
      current_value: null,
      question: 'Please provide a contact phone number so we can coordinate the delivery.'
    });
  }
  
  // Check 4: Vague location
  if (!extractedData.location || 
      extractedData.location.length < 10 || 
      !hasSpecificAddress(extractedData.location)) {
    issues.push({
      type: 'vague_location',
      field: 'location',
      current_value: extractedData.location,
      question: 'Please provide a specific address or landmark. For example: "123 Main Street" or "Lincoln High School, Room 101"'
    });
  }
  
  return {
    is_complete: issues.length === 0,
    issues: issues,
    completeness_score: calculateCompletenessScore(extractedData, issues)
  };
}

// Generate specific questions for vague items
function generateItemSpecificityQuestion(items) {
  const itemQuestions = items.map(item => {
    const lower = item.toLowerCase();
    
    if (lower.includes('medicine') || lower.includes('medical')) {
      return 'What specific medicine or medical supplies do you need? (e.g., bandages, pain medication, insulin, antibiotics)';
    }
    
    if (lower.includes('food')) {
      return 'What specific food items do you need? (e.g., baby formula, canned goods, rice, protein bars)';
    }
    
    if (lower.includes('supplies')) {
      return 'What specific supplies do you need? Please list the exact items.';
    }
    
    return null;
  }).filter(q => q !== null);
  
  return itemQuestions.length > 0 
    ? itemQuestions.join(' ') 
    : 'Please specify exactly what items you need.';
}

// Check if location is specific enough
function hasSpecificAddress(location) {
  const specificIndicators = [
    /\d+\s+\w+\s+(street|st|avenue|ave|road|rd|blvd|drive|dr|lane|ln)/i,
    /room\s+\d+/i,
    /building\s+\w+/i,
    /\d+\s+\w+/,  // Any number followed by text (likely address)
  ];
  
  return specificIndicators.some(pattern => pattern.test(location));
}

// Calculate completeness score (0-100)
function calculateCompletenessScore(data, issues) {
  const totalFields = 5; // items, quantity, location, contact, priority
  const missingFields = issues.length;
  return Math.round(((totalFields - missingFields) / totalFields) * 100);
}

// Use Claude to generate contextual follow-up questions
export async function generateFollowupQuestions(extractedData, userInput) {
  const completenessCheck = checkCompleteness(extractedData);
  
  if (completenessCheck.is_complete) {
    return null; // No follow-up needed
  }
  
  const FOLLOWUP_PROMPT = `You are a disaster response assistant collecting information. 

Original request: "${userInput}"

Extracted data so far:
${JSON.stringify(extractedData, null, 2)}

Issues identified:
${JSON.stringify(completenessCheck.issues, null, 2)}

Generate a helpful, empathetic follow-up message that asks for the missing information. Be:
1. Compassionate (they're in a disaster situation)
2. Clear and specific
3. Brief (ask all questions in one message)
4. Practical

Return ONLY the follow-up message text, nothing else.`;

  try {
    const message = await anthropic.messages.create({
      model: 'claude-sonnet-4-5-20250929',
      max_tokens: 512,
      messages: [
        {
          role: 'user',
          content: FOLLOWUP_PROMPT
        }
      ]
    });

    return {
      needs_followup: true,
      completeness_score: completenessCheck.completeness_score,
      issues: completenessCheck.issues,
      followup_message: message.content[0].text,
      session_id: generateSessionId()
    };

  } catch (error) {
    console.error('Error generating follow-up:', error);
    // Fallback to simple concatenated questions
    return {
      needs_followup: true,
      completeness_score: completenessCheck.completeness_score,
      issues: completenessCheck.issues,
      followup_message: completenessCheck.issues.map(i => i.question).join('\n\n'),
      session_id: generateSessionId()
    };
  }
}

// Merge follow-up responses with original data
export async function mergeFollowupResponse(originalData, followupInput, sessionId) {
  const MERGE_PROMPT = `You are merging follow-up information with an existing disaster request.

Original extracted data:
${JSON.stringify(originalData, null, 2)}

Follow-up response from user:
"${followupInput}"

Update the original data with the new information. Return ONLY valid JSON with the complete, merged data structure. Keep all original fields and update/add based on the follow-up response.`;

  try {
    const message = await anthropic.messages.create({
      model: 'claude-sonnet-4-5-20250929',
      max_tokens: 1024,
      messages: [
        {
          role: 'user',
          content: MERGE_PROMPT
        }
      ]
    });

    const extractedText = message.content[0].text;
    const jsonMatch = extractedText.match(/\{[\s\S]*\}/);
    
    if (!jsonMatch) {
      throw new Error('No valid JSON found in merge response');
    }

    const mergedData = JSON.parse(jsonMatch[0]);
    
    return {
      ...mergedData,
      timestamp: new Date().toISOString(),
      request_id: originalData.request_id,
      raw_input: `${originalData.raw_input}\n[Follow-up]: ${followupInput}`,
      follow_up_completed: true
    };

  } catch (error) {
    console.error('Error merging follow-up:', error);
    throw error;
  }
}

// Generate session ID for tracking multi-turn conversations
function generateSessionId() {
  return `SESSION-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
}

// Store for active follow-up sessions
const activeSessions = new Map();

export function storeSession(sessionId, data) {
  activeSessions.set(sessionId, {
    ...data,
    created_at: new Date().toISOString(),
    expires_at: new Date(Date.now() + 30 * 60 * 1000).toISOString() // 30 min expiry
  });
}

export function getSession(sessionId) {
  return activeSessions.get(sessionId);
}

export function deleteSession(sessionId) {
  activeSessions.delete(sessionId);
}

export function cleanupExpiredSessions() {
  const now = Date.now();
  for (const [sessionId, session] of activeSessions.entries()) {
    if (new Date(session.expires_at).getTime() < now) {
      activeSessions.delete(sessionId);
    }
  }
}

// Run cleanup every 5 minutes
setInterval(cleanupExpiredSessions, 5 * 60 * 1000);

File: services/followup_service.py
"""
Follow-up Service for AgentAid
Handles intelligent follow-up questions and session management
"""

import uuid
import json
from typing import Dict, Any, Optional, List
from datetime import datetime, timedelta

class FollowupService:
    def __init__(self):
        self.active_sessions = {}
        self.session_timeout = 30  # minutes

    def check_completeness(self, extracted_data: Dict[str, Any]) -> Dict[str, Any]:
        """Check if extracted data is complete and specific enough"""
        issues = []

        # Check 1: Vague items
        vague_items = ['medicine', 'medical', 'supplies', 'food', 'items', 'stuff', 'things']
        needs_specificity = any(
            any(vague in item.lower() for vague in vague_items)
            for item in extracted_data.get('items', [])
        )

        if needs_specificity:
            issues.append({
                'type': 'vague_items',
                'field': 'items',
                'current_value': extracted_data.get('items', []),
                'question': self._generate_item_specificity_question(extracted_data.get('items', []))
            })

        # Check 2: Missing or vague quantity
        quantity = extracted_data.get('quantity_needed', '')
        if not quantity or quantity.lower() in ['low', 'medium', 'high']:
            issues.append({
                'type': 'vague_quantity',
                'field': 'quantity_needed',
                'current_value': quantity,
                'question': f"Please specify the exact quantity needed for {', '.join(extracted_data.get('items', []))}. For example: '50 units' or '100 bottles'"
            })

        # Check 3: Missing contact information
        if not extracted_data.get('contact'):
            issues.append({
                'type': 'missing_contact',
                'field': 'contact',
                'current_value': None,
                'question': 'Please provide a contact phone number so we can coordinate the delivery.'
            })

        # Check 4: Vague location
        location = extracted_data.get('location', '')
        if not location or len(location) < 10 or not self._has_specific_address(location):
            issues.append({
                'type': 'vague_location',
                'field': 'location',
                'current_value': location,
                'question': 'Please provide a specific address or landmark. For example: "123 Main Street" or "Lincoln High School, Room 101"'
            })

        completeness_score = self._calculate_completeness_score(extracted_data, issues)

        return {
            'is_complete': len(issues) == 0,
            'issues': issues,
            'completeness_score': completeness_score,
            'needs_followup': len(issues) > 0
        }

    def _generate_item_specificity_question(self, items: List[str]) -> str:
        """Generate specific questions for vague items"""
        questions = []

        for item in items:
            lower = item.lower()

            if 'medicine' in lower or 'medical' in lower:
                questions.append('What specific medicine or medical supplies do you need? (e.g., bandages, pain medication, insulin, antibiotics)')
            elif 'food' in lower:
                questions.append('What specific food items do you need? (e.g., baby formula, canned goods, rice, protein bars)')
            elif 'supplies' in lower:
                questions.append('What specific supplies do you need? Please list the exact items.')

        return ' '.join(questions) if questions else 'Please specify exactly what items you need.'

    def _has_specific_address(self, location: str) -> bool:
        """Check if location is specific enough"""
        import re
        specific_indicators = [
            r'\d+\s+\w+\s+(street|st|avenue|ave|road|rd|blvd|drive|dr|lane|ln)',
            r'room\s+\d+',
            r'building\s+\w+',
            r'\d+\s+\w+',  # Any number followed by text
        ]

        return any(re.search(pattern, location, re.IGNORECASE) for pattern in specific_indicators)

    def _calculate_completeness_score(self, data: Dict[str, Any], issues: List[Dict[str, Any]]) -> int:
        """Calculate completeness score (0-100)"""
        total_fields = 5  # items, quantity, location, contact, priority
        missing_fields = len(issues)
        return max(0, round(((total_fields - missing_fields) / total_fields) * 100))

    def generate_followup_questions(self, extracted_data: Dict[str, Any], user_input: str) -> Optional[Dict[str, Any]]:
        """Generate follow-up questions using AI"""
        completeness_check = self.check_completeness(extracted_data)

        if completeness_check['is_complete']:
            return None

        # Generate contextual follow-up message
        followup_message = self._generate_contextual_followup(extracted_data, user_input, completeness_check['issues'])

        return {
            'needs_followup': True,
            'completeness_score': completeness_check['completeness_score'],
            'issues': completeness_check['issues'],
            'followup_message': followup_message,
            'session_id': self._generate_session_id()
        }

    def _generate_contextual_followup(self, extracted_data: Dict[str, Any], user_input: str, issues: List[Dict[str, Any]]) -> str:
        """Generate contextual follow-up message"""
        if not issues:
            return ""

        # Simple rule-based follow-up generation
        messages = []

        for issue in issues:
            if issue['type'] == 'vague_items':
                messages.append("Please specify exactly what items you need. For example: 'bandages and pain medication' instead of 'medical supplies'.")
            elif issue['type'] == 'vague_quantity':
                messages.append("Please provide specific quantities. For example: '50 units' or '100 bottles'.")
            elif issue['type'] == 'missing_contact':
                messages.append("Please provide a contact phone number so we can coordinate delivery.")
            elif issue['type'] == 'vague_location':
                messages.append("Please provide a specific address or landmark where you need help.")

        return " ".join(messages)

    def merge_followup_response(self, original_data: Dict[str, Any], followup_input: str, session_id: str) -> Dict[str, Any]:
        """Merge follow-up response with original data"""
        # Simple merge strategy - append to additional_notes
        merged_data = original_data.copy()

        current_notes = merged_data.get('additional_notes', '')
        if current_notes:
            merged_data['additional_notes'] = f"{current_notes}\nFollow-up: {followup_input}"
        else:
            merged_data['additional_notes'] = f"Follow-up: {followup_input}"

        merged_data.update({
            'timestamp': datetime.now().isoformat(),
            'follow_up_completed': True
        })

        return merged_data

    def store_session(self, session_id: str, data: Dict[str, Any]) -> None:
        """Store session data"""
        self.active_sessions[session_id] = {
            **data,
            'created_at': datetime.now().isoformat(),
            'expires_at': (datetime.now() + timedelta(minutes=self.session_timeout)).isoformat()
        }

    def get_session(self, session_id: str) -> Optional[Dict[str, Any]]:
        """Get session data"""
        session = self.active_sessions.get(session_id)
        if not session:
            return None

        # Check if expired
        expires_at = datetime.fromisoformat(session['expires_at'])
        if datetime.now() > expires_at:
            del self.active_sessions[session_id]
            return None

        return session

    def delete_session(self, session_id: str) -> None:
        """Delete session"""
        self.active_sessions.pop(session_id, None)

    def cleanup_expired_sessions(self) -> None:
        """Clean up expired sessions"""
        now = datetime.now()
        expired_sessions = []

        for session_id, session in self.active_sessions.items():
            expires_at = datetime.fromisoformat(session['expires_at'])
            if now > expires_at:
                expired_sessions.append(session_id)

        for session_id in expired_sessions:
            del self.active_sessions[session_id]

    def _generate_session_id(self) -> str:
        """Generate unique session ID"""
        return f"SESSION-{int(datetime.now().timestamp())}-{uuid.uuid4().hex[:8]}"
